azurerm/resource_arm_azure_firewall_network_rule_collection.go

```diff
@@ -0,0 +1,342 @@
+	firewall.AzureFirewallPropertiesFormat.IPConfigurations = &ipConfigs
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, firewallName, firewall)
+	if err != nil {
+		return fmt.Errorf("Error creating/updating Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	err = future.WaitForCompletion(ctx, client.Client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for creation/update of Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	return nil
+}
+
+func expandArmAzureFirewallNetworkRuleCollection(d *schema.ResourceData) network.AzureFirewallNetworkRuleCollection {
+	name := d.Get("name").(string)
+	action := network.AzureFirewallRCActionType(d.Get("action").(string))
+	rules := expandArmAzureFirewallNetworkRules(d)
+	properties := network.AzureFirewallNetworkRuleCollectionPropertiesFormat{
+		Action: &network.AzureFirewallRCAction{
+			Type: action,
+		},
+		Priority: utils.Int32(int32(d.Get("priority").(int))),
+		Rules:    &rules,
+	}
+	col := network.AzureFirewallNetworkRuleCollection{
+		Name: &name,
+		AzureFirewallNetworkRuleCollectionPropertiesFormat: &properties,
+	}
+	return col
+}
+
+func findArmAzureFirewallNetworkRuleCollectionByName(firewall *network.AzureFirewall, name string) (*network.AzureFirewallNetworkRuleCollection, int, bool) {
+	for i, collection := range *firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections {
+		if collection.Name != nil && *collection.Name == name {
+			return &collection, i, true
+		}
+	}
+	return nil, -1, false
+}
+
+func removeArmAzureFirewallNetworkRuleCollectionByName(firewall *network.AzureFirewall, name string) *[]network.AzureFirewallNetworkRuleCollection {
+	collections := *firewall.AzureFirewallPropertiesFormat.NetworkRuleCollections
+	for i, collection := range collections {
+		if collection.Name != nil && *collection.Name == name {
+			collections = append(collections[:i], collections[i+1:]...)
+			continue
+		}
+	}
+	return &collections
+}
+
+func expandArmAzureFirewallNetworkRules(d *schema.ResourceData) []network.AzureFirewallNetworkRule {
+	nwRules := d.Get("rule").(*schema.Set).List()
+	rules := make([]network.AzureFirewallNetworkRule, 0)
+
+	for _, nwRule := range nwRules {
+		rule := nwRule.(map[string]interface{})
+
+		name := rule["name"].(string)
+		description := rule["description"].(string)
+		sourceAddresses := rule["source_addresses"].(*schema.Set)
+		destinationAddresses := rule["destination_addresses"].(*schema.Set)
+		destinationPorts := rule["destination_ports"].(*schema.Set)
+		protocols := rule["protocols"].(*schema.Set)
+		ruleToAdd := network.AzureFirewallNetworkRule{
+			Name: &name,
+		}
+		ruleToAdd.Description = &description
+		ruleToAdd.SourceAddresses = expandArmAzureFirewallSet(sourceAddresses)
+		ruleToAdd.DestinationAddresses = expandArmAzureFirewallSet(destinationAddresses)
+		ruleToAdd.DestinationPorts = expandArmAzureFirewallSet(destinationPorts)
```

Comment:
_minor_ can we in-line these? (also we can use `utils.String(val)` rather than `&` for consistency with the other resources) e.g.

```
ruleToAdd := network.AzureFirewallNetworkRule{
  Name: &name,
  Description: utils.String(description),
  SourceAddresses: expandArmAzureFirewallSet(sourceAddresses),
  DestinationAddresses: expandArmAzureFirewallSet(destinationAddresses),
  DestinationPorts: expandArmAzureFirewallSet(destinationPorts),
}
```